core/src/test/java/org/apache/calcite/rex/RexExecutorTest.java

```diff
@@ -378,4 +383,140 @@ public void run() {
   interface Action {
     void check(RexBuilder rexBuilder, RexExecutorImpl executor);
   }
+
+  /**
+   * ArrayList-based DataContext to check Rex execution.
+   */
+  public static class TestDataContext extends SingleValueDataContext {
+    private TestDataContext(Object[] values) {
+      super("inputRecord", values);
+    }
+  }
+
+  /**
+   * Context that holds a value for a particular context name.
+   */
+  static class SingleValueDataContext implements DataContext {
+    private final String name;
+    private final Object value;
+
+    SingleValueDataContext(String name, Object value) {
+      this.name = name;
+      this.value = value;
+    }
+
+    public SchemaPlus getRootSchema() {
+      throw new RuntimeException("Unsupported");
+    }
+
+    public JavaTypeFactory getTypeFactory() {
+      throw new RuntimeException("Unsupported");
+    }
+
+    public QueryProvider getQueryProvider() {
+      throw new RuntimeException("Unsupported");
+    }
+
+    public Object get(String name) {
+      if (this.name.equals(name)) {
+        return value;
+      } else {
+        throw new RuntimeException("Wrong DataContext access");
+      }
+    }
+  }
+
+  /**
+   * User defined executable.
+   */
+  static class UserExecutable implements RexExecutable {
+    final RexBuilder rexBuilder;
+    final List<RexNode> exps;
+    DataContext dataContext;
+    UserExecutable(final RexBuilder rexBuilder, final List<RexNode> exps) {
+      assert exps != null && exps.size() == 1;
+      this.rexBuilder = rexBuilder;
+      this.exps = exps;
+    }
+    @Override public void reduce(RexBuilder rexBuilder, List<RexNode> constExps,
+        List<RexNode> reducedValues) {
+      Object[] values;
+      try {
+        values = execute();
+        assert values.length == constExps.size();
+        final List<Object> valueList = Arrays.asList(values);
+        for (Pair<RexNode, Object> value : Pair.zip(constExps, valueList)) {
+          reducedValues.add(
+              rexBuilder.makeLiteral(value.right, value.left.getType(), true));
+        }
+      } catch (RuntimeException e) {
+        // One or more of the expressions failed.
+        // Don't reduce any of the expressions.
+        reducedValues.addAll(constExps);
+      }
```

Comment:
I think we should call reducedValues.clear before reducedValues.addAll according to your comment.